astropy/modeling/bounding_box.py

```diff
@@ -0,0 +1,573 @@
+        """Validate the shape of an interval representation"""
+        MESSAGE = """An interval must be some sort of sequence of length 2"""
+
+        try:
+            shape = np.shape(interval)
+        except TypeError:
+            try:
+                # np.shape does not work with lists of Quantities
+                if len(interval) == 1:
+                    interval = interval[0]
+                shape = np.shape([b.to_value() for b in interval])
+            except (ValueError, TypeError, AttributeError):
+                raise ValueError(MESSAGE)
+
+        valid_shape = shape in ((2,), (1, 2), (2, 0))
+        if not valid_shape:
+            valid_shape = (len(shape) > 0) and (shape[0] == 2) and \
+                all(isinstance(b, np.ndarray) for b in interval)
+
+        if not isiterable(interval) or not valid_shape:
+            raise ValueError(MESSAGE)
+
+    @classmethod
+    def _validate_bounds(cls, lower, upper):
+        """Validate the bounds are reasonable and construct an interval from them."""
+        if (np.asanyarray(lower) > np.asanyarray(upper)).all():
+            warnings.warn(f"Invalid interval: upper bound {upper} "
+                          f"is strictly less than lower bound {lower}.", RuntimeWarning)
+
+        return cls(lower, upper)
+
+    @classmethod
+    def validate(cls, interval):
+        """
+        Construct and validate an interval
+
+        Parameters
+        ----------
+        interval : iterable
+            A representation of the interval.
+
+        Returns
+        -------
+        A validated interval.
+        """
+        cls._validate_shape(interval)
+
+        if len(interval) == 1:
+            interval = tuple(interval[0])
+        else:
+            interval = tuple(interval)
+
+        return cls._validate_bounds(interval[0], interval[1])
+
+    def outside(self, _input: np.ndarray):
+        """
+        Parameters
+        ----------
+        _input : np.ndarray
+            The evaluation input in the form of an array.
+
+        Returns
+        -------
+        Boolean array indicating which parts of _input are outside the interval:
+            True  -> position outside interval
+            False -> position inside  interval
+        """
+        return np.logical_or(_input < self.lower, _input > self.upper)
+
+    def domain(self, resolution):
+        return np.arange(self.lower, self.upper + resolution, resolution)
+
+
+class BoundingBox(object):
+    """
+    A model's bounding box
+
+    Parameters
+    ----------
+    intervals : dict
+        A dictionary containing all the intervals for each model input
+            keys   -> input index
+            values -> interval for that index
+
+    _model : astropy.modeling.Model
+        The Model this bounding_box is for.
+
+    order : optional, str
+        The ordering that is assumed for the tuple representation of this
+        bounding_box.
+
+    Methods
+    -------
+    validate :
```

Comment:
`validate` can be used to construct a valid bounding box from the old representation.

Reply:
See 9ca6b36.

Reply:
This sort of thing has two purposes, validation and massaging the input into a standard representation. It almost seems like it should have its own name or terminology since it isn't simply validation. Validate and regularize or something like that. This is more of a pointless comment that is more for future terminology than in this case.